### ISSUE TITLE
Detect variables which may be undefined in some code paths

### ISSUE THREAD

**Issue by brettcannon** [Microsoft/python-language-server#299]
Microsoft/vscode-python#2958

Try to help people detect `UnboundLocalError`\`NameError` situations like:
```python
def example():
    if True:
        pass
    else:
        x = 42
    print(x * 2)

example()
```
or
```python
def example():
    print(x * 2)
    x = 42

example()
```
(Make the above examples global-level statements and they become `NameError`.)

**Comment by MikhailArkhipov** [comment]
I'd

**Comment by MikhailArkhipov** [comment]
@jakebailey - wrt the message wording

**Comment by qubitron** [comment]
Need to move messages to resources
Remove (analysis), (parser) etc.

**Comment by jakebailey** [comment]
I think that first one could actually just be solved by #313 (dataflow analysis recognizing a code path without the variable being defined), so I'll do the naming fixes then close this when those are in.
